### DISCUSSION TITLE
Descriptor retrieval

### DISCUSSION THREAD

**Discussion by EvanKomp** [rvianello/chemicalite#10] [Q&A]
Is there an effective way to retrieve binary fingerprint objects in a vector form?

I have created index tables for querying using fingerprints as was outlined in the tutorials -  but now it would be nice to be able to retrieve those fingerprints for external use to eg. create a graphic of chemical space of different subsets of the database.

I (with total C++ ignorance) saw `bfp_to_blob` in `bfp.cpp` so I attempted something along the lines of `SELECT bfp_to_blob(fp) FROM morgan_idx_molecule` hoping to convert the bfps in the `fp` column to a blob such that I might be able glean a vector from it.

So: is there a clean way to retrieve the vectors associated with fingerprints stored in the db?

Thanks for any help,
Evan

**Comment by rvianello** [comment]
Hi @EvanKomp 
molecules and binary fingerprints are stored as binary blobs that basically consist of their serialized data, just prepended by a 4-bytes header, so I think what you'd like to have is actually more similar to `blob_to_bfp`. Unfortunately, this function is not currently directly available as part of the exposed SQL api, but in the meantime you could maybe try removing the header from the binary data returned by the select statement?

This is quite straightforward to do for molecule data:

    In [69]: mol_blob = cursor.execute("""SELECT mol_from_smiles('Cc1ncccc1')""").fetchone()[0]
    
    In [70]: mol = Chem.Mol(mol_blob[4:])
    
    In [71]: print(Chem.MolToSmiles(mol))
    Cc1ccccn1

And getting access to the bfp binary data should be equally simple

    In [72]: bfp_blob = cursor.execute(
        ...:     """SELECT mol_morgan_bfp(mol_from_smiles('Cc1ncccc1'), 2, 1024)"""
        ...: ).fetchone()[0]
    
    In [73]: bfp_data = bfp_blob[4:]
    
    In [74]: type(bfp_data), len(bfp_data)
    Out[74]: (bytes, 128)

but in some tests I performed this morning I couldn't find an obvious way to for example instantiate an RDKit `ExplicitBitVect` from the raw `bytes` data (maybe I'm missing something, or the exposed python API expects a different serialization format). I'll further investigate this, but how would you more specifically plan to process the binary fingerprints? Would the raw binary vector help, and if not, how would you expect the bfp data to be available?
